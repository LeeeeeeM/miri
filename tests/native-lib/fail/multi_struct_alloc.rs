//@compile-flags: -Zmiri-permissive-provenance

#[repr(C)]
#[derive(Copy, Clone)]
struct HasPointer {
    ptr: *const u8,
}

extern "C" {
    fn access_struct_ptr(s: HasPointer) -> u8;
}

fn main() {
    let vals = [10u8, 20u8];
    let structs =
        vec![HasPointer { ptr: &raw const vals[0] }, HasPointer { ptr: &raw const vals[1] }];
    unsafe {
        access_struct_ptr(structs[1]);
        let _val = *std::ptr::with_exposed_provenance::<u8>(structs[0].ptr.addr()); //~ ERROR: Undefined Behavior: attempting a read access using <wildcard>
    };
}
