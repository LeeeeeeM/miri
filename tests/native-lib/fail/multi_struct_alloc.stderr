error: Undefined Behavior: attempting a read access using <wildcard> at ALLOC[0x0], but no exposed tags have suitable permission in the borrow stack for this location
  --> tests/native-lib/fail/multi_struct_alloc.rs:LL:CC
   |
LL | ...al = *std::ptr::with_exposed_provenance::<u8>(structs[0].ptr.addr());
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ this error occurs as part of an access at ALLOC[0x0..0x1]
   |
   = help: this indicates a potential bug in the program: it performed an invalid operation, but the Stacked Borrows rules it violated are still experimental
   = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
   = note: BACKTRACE:
   = note: inside `main` at tests/native-lib/fail/multi_struct_alloc.rs:LL:CC

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to 1 previous error

